annotation-processing/src/main/java/com/exactpro/epfast/annotation/processing/FastProcessor.java

```diff
@@ -9,122 +12,253 @@
+                errorReporter.reportServiceUpdate(SERVICE_TYPENAME);
             }
         }
+
         if (annotations.contains(fastTypeElement)) {
-            if (fastTypesByName != null) {
-                processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR,
-                    "@FastType annotation processor can not handle classes with @FastType " +
-                        "annotation added after first processing round. " +
-                        "It may be caused by activity of other annotation processors.");
+            if (fastCreator != null) {
+                errorReporter.reportProcessingAfterFirstRound();
             }
-            fastTypesByName = collectFastTypes(roundEnv);
+            fastPackageResolver = resolveFastTypes(roundEnv);
+            fastCreator = resolvePackages(fastPackageResolver);
+            HashMap<String, HashSet<Element>> allPackages = getAllFastPackages(roundEnv);
+            removeAndReportInvalidPackages(allPackages, fastCreator);
+            removeAndReportDuplicates(allPackages, fastCreator);
+            fastCreator.addEmptyPackages(allPackages.keySet());
+            Collection<AbstractFastPackage> fastPackages = fastCreator.getAllFastPackages();
+            fastPackages.removeIf(this::containsDuplicateFastTypes);
+            removeAndReportIfRootEqualsFastPackage(fastCreator);
             try {
-                buildCreatorClass(fastTypesByName);
-            } catch (Exception e) {
-                processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR, e.getMessage());
+                buildCreatorClass(fastCreator);
+            } catch (IOException e) {
+                errorReporter.reportIOException(e);
             }
             return true;
         }
-
         return false;
     }
 
-    private HashMap<String, Element> collectFastTypes(RoundEnvironment roundEnv) {
-        HashMap<String, ArrayList<Element>> nameClassMap = new HashMap<>();
+    private void removeAndReportIfRootEqualsFastPackage(FastCreator fastCreator) {
+        AbstractFastPackage rootPackage = fastCreator.getRootPackage();
+        if (rootPackage == null) {
+            return;
+        }
+        Collection<AbstractFastPackage> fastPackages = fastCreator.getAllFastPackages();
+        long duplicates = fastPackages.stream().filter(
+            fastPackage -> fastPackage.getName().equals(rootPackage.getName()) && !fastPackage.isRootPackage()).count();
+        if (duplicates > 0) {
+            fastCreator.removePackage(rootPackage.getName());
+            errorReporter.reportRootEqualsFastPackage(rootPackage.getName());
+        }
+    }
+
+    private boolean containsDuplicateFastTypes(AbstractFastPackage fastPackage) {
+        List<Element> fastTypes = fastPackage.getFastTypes();
+        HashMap<String, ArrayList<Element>> fastTypeNames = collectFastTypeNames(fastTypes);
+        AtomicBoolean invalid = new AtomicBoolean(false);
```

Comment:
You don't perform any parallel processing here. You can you just regular boolean type